Try using variable for root padding value

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

* Adds a setting to `theme.json` so themes can opt in to the new behaviour of root padding
* Adds a default non-zero value for root padding
* Outputs that value as a custom property on the root element
* Allows full-width blocks to compensate for that value with negative margins

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Partly fixes #35607 - only addresses the root block part of the equation. #39871 deals with full width in nested blocks.
Fixes #35884 for themes that opt into this change. 

Themes shouldn't have to add their own rules to make full width alignments work properly.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

First, we're setting a default padding value in theme.json. It can be overridden by any user-set padding in the global styles interface.

Then, on the server side, we set top and bottom padding to the root container, but left and right padding to the direct children of root. This is so that those children still stretch full width as expected, but _their_ children get the effects of the padding.

On the site editor side, the direct children of root container get the left/right padding applied, so they still appear as full-width. (If we wanted to only apply full-width to some of the root children, maybe we could add padding to the root and enable the align control on them so they could be set manually? And add negative margins to `alignfull` blocks)

On the post editor side, the root container also gets its left/right padding, and any `alignfull` children are fixed with negative margins.


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

0. Add `"useRootVariables": true` to the `settings` of your theme's `theme.json` to enable these changes.
1. Add a few layout blocks (e.g. Group) to the post and the site editor
2. Make some of those blocks full width
3. Check that there's a default padding value in both editors and front end, and full width blocks behave as expected
4. Change root padding value in the site editor by going to Global Styles > Layout > Padding
5. Check that new padding value works in both editors and front end.
6. Check that removing `"useRootVariables": true` from `theme.json` reverts to the previous behaviour.

## Screenshots or screencast <!-- if applicable -->
![ezgif com-gif-maker (5)](https://user-images.githubusercontent.com/8096000/160996605-fcc4d868-1c06-434d-85dd-a94f4b29bb1f.gif)

tellthemachines committed:
Try using variable for root padding value

tellthemachines commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -16,6 +16,46 @@
+		'--wp--style--padding-top'    => array( 'spacing', 'padding', 'top' ),
+		'--wp--style--padding-right'  => array( 'spacing', 'padding', 'right' ),
+		'--wp--style--padding-bottom' => array( 'spacing', 'padding', 'bottom' ),
+		'--wp--style--padding-left'   => array( 'spacing', 'padding', 'left' ),
```
I'm not sure why this only seems to affect the root styles 🤔

tellthemachines commented on lib/compat/wordpress-6.0/theme.json:
```diff
@@ -0,0 +1,253 @@
+				"right": "17px",
+				"bottom": "17px",
+				"left": "17px"
+			}  
```
The "padding" object is the only real addition to this file.

youknowriad replied:
Is this like a demo padding that applies to all themes? won't this be a breaking change for existing themes?

tellthemachines replied:
Yeah, I guess it could be if the theme doesn't define any root padding. I was mainly thinking of a fix for #35884, but perhaps that should be post editor specific.

scruffian replied:
Yeah this is problem for themes that aren't opting in to site padding. I think we'll need to add it programatically.

tellthemachines commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -319,6 +319,11 @@
 
+.is-root-container > .alignfull {
+	margin-right: calc(var(--wp--style--padding-right) * -1);
+	margin-left: calc(var(--wp--style--padding-left) * -1);
```
Not sure if we should have a fallback value here in case the variable doesn't exist.

tellthemachines commented on packages/blocks/src/api/constants.js:
```diff
@@ -118,6 +118,26 @@ export const __EXPERIMENTAL_STYLE_PROPERTY = {
+	'--wp--style--padding-top': {
+		value: [ 'spacing', 'padding', 'top' ],
+		support: [ 'spacing', 'padding' ],
+		rootOnly: true,
```
This feels super hacky, but without it these styles get applied to all blocks that have padding set on them. Any ideas welcome!

ramonjd replied:
Should global styles also output these vars for specific blocks?

I bumped the padding value of the Group block in global styles and I see this in the `global-styles-inline-css` style tag:

```css
body {
	...
	--wp--style--padding-top: 173px;
	--wp--style--padding-right: 173px;
	--wp--style--padding-bottom: 173px;
	--wp--style--padding-left: 173px;
	--wp--style--block-gap: 1.5rem;
}
...
.wp-block-group {
	--wp--style--padding-top: 142px;
	--wp--style--padding-right: 142px;
	--wp--style--padding-bottom: 142px;
	--wp--style--padding-left: 142px;
}
```

The editor reflects the changes but the frontend doesn't, because, I guess, we'd need to either update the blocks' CSS to use these vars or print it on the fly somewhere in the backend.

On a related note, gap also gets special treatment in [WP_Theme_JSON_Gutenberg](https://github.com/WordPress/gutenberg/blob/848db705d6436bc267234910eb653968e6bf2893/lib/compat/wordpress-5.9/class-wp-theme-json-5-9.php#L303) via  `'blockGap' => 'top',`, the effect of which is that `--wp--style--block-gap` is only rendered at the "top" (or root) level in global styles. Effectively the body tag.

But we also want block-level global styles, when set, to override any root value. It's something we're trying resolve over at https://github.com/WordPress/gutenberg/pull/39870

It seems like we do need some sort of flag that says _"Hey, this property only lives at the root level, so output its style rule in the body tag and nowhere else"_.

I'm not sure what that would look like yet, maybe a `const VALID_TOP_LEVEL_STYLES` in `WP_Theme_JSON_Gutenberg` that explicitly defines root-level-only styles 🤷

tellthemachines replied:
> Should global styles also output these vars for specific blocks?

Oooooh no it shouldn't! It's actually outputting the variables instead of the properties in the front end now, and that's messing with some of the negative margin calculations 😫 

> It seems like we do need some sort of flag that says "Hey, this property only lives at the root level, so output its style rule in the body tag and nowhere else".

Yup, we definitely need something like this. I was playing around with it on the editor side here, not realising it would also be needed on the server side.

> I'm not sure what that would look like yet, maybe a const VALID_TOP_LEVEL_STYLES in WP_Theme_JSON_Gutenberg that explicitly defines root-level-only styles

Can we make do with the "top" value in `VALID_STYLES` solution we have currently? What shortcomings do you see with it?

tellthemachines replied:
Ok I made some changes that should fix the block-level global styles issue. 

The thing with padding is that we only need it to be a variable at the root level, so we can easily apply it to all the places we need it: as padding on full-width root level blocks, or as negative margin on their full-width children. I'm not sure if block gap works exactly like that? (I only read through the code of #39870 today, but will look at it more in-depth on Monday).

ramonjd replied:
> The thing with padding is that we only need it to be a variable at the root level, so we can easily apply it to all the places we need it: as padding on full-width root level blocks, or as negative margin on their full-width children. I'm not sure if block gap works exactly like that? 

Block gap is a smidge different, for sure.

Support for gap in non root-level global styles was removed, and in `sanitized()` it's being excluded from the non-root level blocks and elements in the merged theme json.  That means that, right now, there's no need to worry about whether the styles will be correctly output.

The side-effect of that removal was that block-level global styles cannot define gap values; this side-effect was, I believe, unintended. This is the "bug" that https://github.com/WordPress/gutenberg/pull/39870 is trying to address.

This is a long-winded way of saying that I'd like to take block gap in the same direction you're taking padding 😄 so I _think_ there might be similarities...

jasmussen replied:
Conceptually comparing the root padding to block gap makes sense for the reasons Isabel outlines. Just as an example, we'll need to apply this root padding inside the modal navigation overlay as well, so the overlay padding is the same as the root padding.

tellthemachines replied:
Oh, thanks for the reminder @jasmussen ! I'd forgotten about the overlay special case. Do you know of anywhere else where we might need to apply root padding?

scruffian replied:
There are many cases where we might want the site padding to match other parts of the site - for example the gaps between navigation items, template parts and galleries etc. This is an important consideration in theme design - if these spaces all defaulted to something that matched it would mean that designs would look good without needing to tweak all these things individually as we do today cc @kjellr @beafialho

tellthemachines replied:
> There are many cases where we might want the site padding to match other parts of the site - for example the gaps between navigation items, template parts and galleries etc.

That would be a theme level decision though, wouldn't it? I'd imagine the theme would set a global spacing variable and use it to set both root and block padding (or gap, or whatever else) in theme.json and in the theme stylesheet.

The root padding we're dealing with here is a [top-level style](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#top-level-styles) so it's only attached to the body. This PR changes the output from a straightforward padding declaration to a custom property, mainly so we can use its value to offset full-width blocks.

scruffian replied:
Yes, that's a theme level decision, but I think its the kind of thing that having a good default for would really help - for example if this setting defaulted to `--wp--style--block-gap` then all these spacing elements would match by default.

tellthemachines replied:
Oh, I see! 

I changed the default values [here](https://github.com/WordPress/gutenberg/blob/044f281ff5c4907fdcbfeea5d9225e2962e9aada/lib/compat/wordpress-6.0/theme.json#L245) to use the block gap variable - does that work better?

scruffian replied:
Yeah that look good!

tellthemachines commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -190,6 +191,10 @@ function getStylesDeclarations( blockStyles = {} ) {
 
+	if ( isRoot ) {
+		return output;
+	}
```
This is a total hack 😅 to avoid the output of actual padding properties by the style engine logic below. Might be better to work out a way of getting the style engine to output the variables instead.

andrewserong replied:
I'm fine with the hack here, personally — it shouldn't be too hard to migrate the logic over to a separate file in the style engine package after the fact in a follow-up if you don't want to worry about it in this PR.

jasmussen commented:
Nice one, thanks so much for taking this on, it's an important quality of life fix for block themes!

Question: does it only work with root level full-wide blocks, though? I can't seem to get it to work in this featured image use case, where the featured image is part of post content:
![root padding](https://user-images.githubusercontent.com/1204802/161002338-fb2039de-8077-4a76-852c-cfb12ad26c3b.gif)

MaggieCabrera commented:
/cc @WordPress/block-themers

ramonjd commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -334,9 +339,10 @@ export const toStyles = ( tree, blockSelectors ) => {
 	nodesWithStyles.forEach( ( { selector, styles } ) => {
-		const declarations = getStylesDeclarations( styles );
+		const isRoot = ROOT_BLOCK_SELECTOR === selector;
+		const declarations = getStylesDeclarations( styles, isRoot );
```
I'm doing something similar in the backend over at https://github.com/WordPress/gutenberg/pull/39870/files#diff-c220112e642d42482d91af8a96dca5ab587adfa705140148f86c064bca995631R184, but passing the whole selector.

tellthemachines commented:
Thanks for testing this @jasmussen! I'm not sure I fully get what's going on in that gif 😅  How does the Query Loop go inside the Post Content? Could you share some basic steps to set up that scenario?

> does it only work with root level full-wide blocks, though?

This PR only addresses root level padding, and its effects on full width blocks that are direct children either of `wp-site-blocks` or of the post content. There's a separate effort over at #39871 for adjustments to padding/full width in nested blocks (where full width is expected to be relative to a container block and not the viewport). I should have mentioned that in the PR description 🤦

tellthemachines committed:
Root vars shouldn't apply to block global styles

tellthemachines commented:
One thing that's not working quite as expected here is the top/bottom padding values added to `editor-styles-wrapper` in the site editor: when changing the root padding in global styles, they only update on page refresh. I'm pretty sure the rule adding those styles is coming from the block editor settings, but I haven't found where it gets added to the settings yet. Any tips on that appreciated!

jasmussen commented:
> Thanks for testing this @jasmussen! I'm not sure I fully get what's going on in that gif 😅 How does the Query Loop go inside the Post Content? Could you share some basic steps to set up that scenario?

I haven't been following too closely recent changes to post content or query loop, but here's a GIF walking through activating the empty theme and then selecting the singular template and seeing the issue:

![steps](https://user-images.githubusercontent.com/1204802/161206843-8e4a3d69-dab2-43ee-bd1a-97e1013a7644.gif)

But I see the issue even if I drag the Query loop outside of the Post Content block:

<img width="1392" alt="Screenshot 2022-04-01 at 08 22 33" src="https://user-images.githubusercontent.com/1204802/161206917-371afb29-a80e-4d35-ade4-8a24f4654303.png">

> This PR only addresses root level padding, and its effects on full width blocks that are direct children either of wp-site-blocks or of the post content. There's a separate effort over at https://github.com/WordPress/gutenberg/pull/39871 for adjustments to padding/full width in nested blocks (where full width is expected to be relative to a container block and not the viewport). I should have mentioned that in the PR description 🤦

Oh! My mistake. Ignore everything I said above 🙈 

Thanks for working on this!

ramonjd commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -16,6 +16,50 @@
+		'margin-right'                      => array( 'spacing', 'margin', 'right' ),
+		'margin-bottom'                     => array( 'spacing', 'margin', 'bottom' ),
+		'margin-left'                       => array( 'spacing', 'margin', 'left' ),
+		'padding'                           => array( 'spacing', 'padding' ),
```
Should we give the same treatment to 'padding'?

If I add a padding value to theme.json, 

```json
                "spacing": {
                        "blockGap": "1.5rem",
                        "padding": "88px"
                },
```

I see the following:

```css
body {
	padding: 88px;
	--wp--style--block-gap: 1.5rem;
}
```

Not sure what the intention is here, but If there were another CSS var for a single padding value, e.g., `--wp--style--root--padding`, we could use it as a fallback, e.g., 

```css
padding-top: var(--wp--style--root--padding-top, var(--wp--style--root--padding))
```

tellthemachines replied:
Oh, good point, if the theme just sets "padding" my carefully crafted logic will fail 😂 😭

tellthemachines replied:
We won't be able to use `--wp--style--root--padding` directly as a fallback though, because it might be a multi-value format e.g. `32px 42px`. Not quite sure how best to handle that - is that something that we've solved for gap already?

ramonjd replied:
> We won't be able to use --wp--style--root--padding directly as a fallback though, because it might be a multi-value format e.g. 32px 42px. 

Oh, excellent point. 😆 Good catch.

I'm not sure splitting by `' '` would be robust enough (I tried in an old gap commit and was told the same). 

As for gap, the gap value can be an array, so we check for that and split the values accordingly. 

See: https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L70

Otherwise we assume string.

tellthemachines replied:
Hmmm, here we'll always have a string so that won't work. 

If we could identify which of the variables are set, we could maybe output different styles accordingly, but I'm not sure how that would work on the editor side (the whole editor side of this PR is still quite iffy though, so maybe a solution will present itself in time)

tellthemachines replied:
Thinking better of this, we always need to know the specific values for left and right padding because we need them for the negative margins on full width blocks. For that we need to separate the multi-value padding string into its component parts, so I think we'll have to do it the messy way.

tellthemachines replied:
I updated the code below to separate out the values from a shorthand padding declaration. This currently only works on the server side; looking into how to implement it for the editor now.

ramonjd commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -142,14 +142,18 @@ function flattenTree( input = {}, prefix, token ) {
  * @return {Array} An array of style declarations.
  */
-function getStylesDeclarations( blockStyles = {} ) {
+function getStylesDeclarations( blockStyles = {}, isRoot = false ) {
```
Ignorable observation: would it make sense to keep the selector arg consistent with how we're doing in the `compute_style_properties()` PHP method?

That is, passing the `selector` and doing the check inside this method?

```js
ROOT_BLOCK_SELECTOR === selector
```

ramonjd commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -119,6 +166,81 @@ protected function get_block_classes( $style_nodes ) {
+		}
+
+		// If a variable value is added to the root, the corresponding property should be removed.
+		foreach ( $root_variable_duplicates as $duplicate ) {
```
I've been playing with an alternative for [block gap](https://github.com/WordPress/gutenberg/pull/39870), in which I've created a second array `const ROOT_PROPERTIES_METADATA` that contains all the possible root properties.

In this case it would look like this:

```php
const ROOT_PROPERTIES_METADATA = array(
    // all the rest here...
    // except "padding", "padding-top" etc, which are replaced by the following CSS vars
    '--wp--style--root--padding' => array( 'spacing', 'padding' ),
    '--wp--style--root--padding-top' => array( 'spacing', 'padding', 'top' ),
    '--wp--style--root--padding-right' => array( 'spacing', 'padding', 'right' ),
    '--wp--style--root--padding-bottom' => array( 'spacing', 'padding', 'bottom' ),
    '--wp--style--root--padding-left' => array( 'spacing', 'padding', 'left' ),
);
```

Then the only change to `compute_style_properties()` method would be:

```php
f ( null === $properties ) {
    $properties =  static::ROOT_BLOCK_SELECTOR === $selector ? static::ROOT_PROPERTIES_METADATA : static::PROPERTIES_METADATA;
}
```

Just noting for reference! It might be completely bonkers.

Seems to work for padding at least 😆 but haven't battle tested it yet. 🤔

tellthemachines replied:
Ooooh that's a great idea! Much neater logic.

tellthemachines committed:
Fix issue with shorthand padding on server side.

tellthemachines committed:
Temp fix for file path error

tellthemachines commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -319,6 +319,11 @@
+.is-root-container > .alignfull {
+	margin-right: calc(var(--wp--style--root--padding-right) * -1);
+	margin-left: calc(var(--wp--style--root--padding-left) * -1);
+}
```
This is where it gets really tricky to handle shorthand padding values. Perhaps we can add this rule in dynamically somewhere else.

youknowriad replied:
If I'm not wrong it's not possible to have "align full" in the site editor (or any editor) unless there's a layout defined in the root which only happens in the post editor if the theme.json has "layout" key right? Makes me wonder if this should be moved to the edit-post or something like that to explain its purpose (maybe add a comment)

jasmussen commented:
> Do you know of anywhere else where we might need to apply root padding?

Nothing comes to mind, but I would note that I could see the overlay having a variety of behaviors some time the future, such as being an overlay that opens from the right and scrims the content, or a modal in the center. So it might not always need root padding. That is, it's probably fine to set it by default, and then let the user customize that padding on the modal overlay directly — and definitely a future concern, nothing for this PR, just context.

tellthemachines committed:
Fix site editor top padding not updating

tellthemachines committed:
Support shorthand properties in site editor.

tellthemachines committed:
Fix full widths in the post editor

youknowriad commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -174,6 +178,54 @@ function getStylesDeclarations( blockStyles = {} ) {
+			} else if ( !! properties && isString( styleValue ) && rootOnly ) {
+				const separateValues = styleValue.split( ' ' );
+
+				const sortedBoxValues = {
```
What is this default value about?

tellthemachines replied:
We should probably decide on a fallback value if neither theme nor user provides one; I just used 17px throughout this PR to make debugging easier. But here, I guess it could be 0 or even null, because we always fill in a value for all the properties below.

youknowriad reviewed (commented):
Here are a few things I noticed. 

I'm using a full aligned image inside a group block with a layout defined.

 - If I put my group block at the root level, everything works as expected on the frontend 
 - In the site editor root blocks I don't see any padding around random root blocks  (not sure why).
 - If I make my group block a template part, the full width images are not stretched anymore in the frontend.

In my singular template, I have a "post-content" block with layout but it's inside a group block to be able to apply some styling (background... and add some content after it...)

When I do that, full width aligned images in a post are not full width in the frontend, I can see the root padding around them.

tellthemachines committed:
Check selector inside function.

tellthemachines committed:
Fix kebab-casing of CSS variables.

tellthemachines commented:
Thanks for testing @youknowriad !

> In the site editor root blocks I don't see any padding around random root blocks (not sure why).

That was a bug with variable names being kebab-cased, I've fixed it now.

> If I make my group block a template part, the full width images are not stretched anymore in the frontend.

This PR is only dealing with full-aligned blocks that are direct children of root blocks. When the group is wrapped in a template part, it's no longer a root block so the styles don't apply anymore.Non-root full width layouts are being addressed in #39871, as they require a different logic (I think; at least I haven't found a way to address both simultaneously 😅 )

> When I do that, full width aligned images in a post are not full width in the frontend, I can see the root padding around them.

That's the same problem as above with the nesting. I'm hoping that we might also be able to address that scenario in #39871. Will look into it better tomorrow!

scruffian commented:
One note for people testing this - almost every theme already contains CSS to control this, so you will likely see some conflicts between that CSS and this PR unless you test with emptytheme. That's not to say we shouldn't test with other themes - we need to, to check that there aren't any regressions...

tellthemachines committed:
Fix borked merge conflict solving

tellthemachines committed:
Move post editor full width styles to edit-post package.

tellthemachines committed:
Set default padding value to 0.

tellthemachines committed:
Update test string.

tellthemachines committed:
Fix PHP unit tests

tellthemachines committed:
Fix PHP lint

tellthemachines commented:
Thanks everyone for your input so far! I've addressed most of the issues that came up, and I think there's mostly one question remaining: Will setting a non-zero root padding default in Core mess with existing themes in any way?

My guess is this might only be an issue for themes that _don't_ set any kind of root padding, whether through `theme.json` or not, but it would be great to have some input on this question from more theme-savvy folks 🙂  Cc. @scruffian @kjellr @MaggieCabrera 

(Setting a non-zero default would fix #35884, which is why I was keen to do it)

Apart from that, more feedback on the code would be very welcome! I haven't worked out how best to interact with the style engine [here](https://github.com/WordPress/gutenberg/pull/39926/files#r839255660), so thoughts welcome there too.

tellthemachines committed:
Fix failing PHP tests.

tellthemachines committed:
Use block gap variable as default root padding value.

andrewserong commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -437,12 +486,157 @@ protected static function get_metadata_boolean( $data, $path, $default = false )
+				$shorthand_bottom = '0';
+				$shorthand_left   = '0';
+		
+				$separate_values = explode( ' ', $value );
```
I was thinking about this `explode()` call and the similar one in JS (`styleValue.split( ' ' );`), and the issue https://github.com/WordPress/gutenberg/issues/40132 — would it be worth it to add a parser function so that we can handle values that contain CSS functions like `calc()` `var()`, etc?

Because of nesting it's probably outside the scope of a simple regex, but I hacked together the following in JS to see if it'd be viable:

```js
const inputString = 'var( --my-variable, calc(2px + 3px) ) 2px 3px 0px';

function parseShorthandProps( input ) {
  const values = [];
  
  let openBracket = 0;
  let closedBracket = 0;
  let startIndex = 0;
  let isCollecting = false;
  
  for ( let i = 0; i < input.length; i++ ) {
    if ( ! isCollecting && input[i] !== ' ' ) {
      isCollecting = true;
      startIndex = i;
    }
    
    if ( input[i] === ('(') ) {
      openBracket++;
    }
    
    if ( input[i] === (')') ) {
      closedBracket++;
    }

    if ( isCollecting && ( input[i] === ' ' && openBracket === closedBracket ) || i === input.length - 1 ) {
      isCollecting = false;
      const val = i === input.length -1 ? input.substring( startIndex ) : input.substring( startIndex, i );
      values.push( val );
      openBracket = 0;
      closedBracket = 0;
    }
  }
  
  return values;
}

console.log( parseShorthandProps( inputString ) );
// output: ["var( --my-variable, calc(2px + 3px) )","2px","3px","0px"]
```

If so, perhaps we could add a separate method for that calculation so it doesn't balloon out this function... I was mostly just curious to see if it's possible, but also totally understand if you think it's overkill / too hacky!

ramonjd replied:
> would it be worth it to add a parser function so that we can handle values that contain CSS functions like calc() var(), etc?

Nice! This would be helpful right across theme.json, where style input values can be any string.  

Something about it makes me wary, and I'm not sure I can articulate it clearly. I think we'd need to define the patterns that we accept and see how these values function across the application. 

For example, how would `"clamp( var(--my-variable), calc(2px + 3px), 3.4em ) 100px var( --wp-admin-theme-color, 10px ) calc(2px + 2vw)"` affect editor UI controls that expect `value + unit`?

I'm not sure what the right answer is here. Theme authors can add any string they like right now without any guarantee that it'll work. 

I guess the difference is that, if we starting parsing these values out, there might be an expectation that they'll work everywhere.

In the context of padding, we have patterns to define sides `{ top, right, bottom, left }`. 

In the first iteration, would it be safer to assume `string === ''{unit}{value}'` and `array() === [ top, right, bottom, left ]`?  

I don't know. Asking for a friend :trollface:

tellthemachines replied:
Oooh nice, thanks for writing that! I don't think it's hacky at all.

I'm on the fence as to whether we should add it though; in fact, even the logic already in here might not be necessary if we decide not to support shorthand values at all. There are already issues with them elsewhere (#40132) so it would be good to make a decision and document it explicitly.

andrewserong replied:
> Something about it makes me wary, and I'm not sure I can articulate it clearly. I think we'd need to define the patterns that we accept and see how these values function across the application.

Thanks for the thoughts, very nicely articulated!

> I guess the difference is that, if we starting parsing these values out, there might be an expectation that they'll work everywhere.

I think that's definitely the challenge here — if we add a parser, it makes the code more tolerant to values that we don't actually support in any editable way, so I'd think of it as a way to massage values into something we can work with, rather than something we officially support. I don't personally see any issue in saying, "We don't support that yet" and defer adding parsing, etc, to such a point as we want to investigate non `value + unit` values in the property. I was mostly curious to see if we could do it if we need to 😊

andrewserong replied:
> There are already issues with them elsewhere (https://github.com/WordPress/gutenberg/issues/40132) so it would be good to make a decision and document it explicitly.

Great point, perhaps it's something to explore in a separate PR. Happy to take your lead on that! I'll copy my code snippet over to that issue for posterity.

andrewserong reviewed (commented):
Nice work, this is testing pretty well for me using `emptytheme` @tellthemachines (in Twenty Twenty Two, an additional padding appears, but I think that's to be expected based on Ben's comment).

> On the site editor side, the direct children of root container get the left/right padding applied, so they still appear as full-width.

I just wanted to double-check that I understand the intended behaviour here. Here's a screenshot of what I observed — basically, this means that the Group block added at the root level is effectively always full-width?

| This PR | Trunk |
| --- | --- |
| <img width="399" alt="image" src="https://user-images.githubusercontent.com/14988353/162658062-46dc5824-5658-49b3-a3b8-e91518fd9820.png"> | <img width="397" alt="image" src="https://user-images.githubusercontent.com/14988353/162658624-4307ce71-55ad-4fe0-8203-ebc5ca565990.png"> |

My understanding is:

* The root padding setting will be applied in global styles to all children of root, so if there is no block-level padding set, then the global setting will apply, and this effectively makes the root blocks full width by default.
* If a root block has padding set, this value will be used instead of the global padding.
* If a user wants the root level padding to visually be applied to each element within their site layout, they need to nest within a root Group block (I imagine this will happen anyway most of the time due to folks using template parts).

It took me a moment to wrap my head around (I think I'm used to the idea that the global padding value sets padding around everything), but now that I've played with it for a bit, I really like the behaviour in this PR as it makes it much easier to have some sections with padding, some without, and by default creating sections that are full-width is much easier.

Since this changes the default behaviour at the root, though, it sounds like it could potentially break existing layouts that expected the root padding to add that additional whitespace surrounding group blocks that already have their own padding value set?

tellthemachines commented:
Thanks for reviewing @andrewserong !

> My understanding is:

>The root padding setting will be applied in global styles to all children of root, so if there is no block-level padding set, then the global setting will apply, and this effectively makes the root blocks full width by default.
>If a root block has padding set, this value will be used instead of the global padding.
>If a user wants the root level padding to visually be applied to each element within their site layout, they need to nest within a root Group block (I imagine this will happen anyway most of the time due to folks using template parts).

Yup, that's correct!

> Since this changes the default behaviour at the root, though, it sounds like it could potentially break existing layouts that expected the root padding to add that additional whitespace surrounding group blocks that already have their own padding value set?

Yes, if a theme or a website has set root padding and is counting on it applying around all root blocks, so that they aren't full width, then this is a breaking change. I'm not sure if that's acceptable, or if we should ensure back compat by making this an optional behaviour. There's some mention of adding a toggle in #35607; I guess that's something we could explore.

tellthemachines committed:
Fix linting errors.

andrewserong commented:
Thanks for confirming @tellthemachines! That makes good sense.

> some mention of adding a toggle

That'd be a neat way to handle it for backwards compatibility if we go down that path. And at least in the global styles UI, I suppose there's plenty of room to add a toggle near Padding without it muddying up the interface:

<img width="300" src="https://user-images.githubusercontent.com/14988353/162673067-8535084e-0ac8-414c-9704-202c0c402820.png" />

beafialho commented:
> There are many cases where we might want the site padding to match other parts of the site - for example the gaps between navigation items, template parts and galleries etc. This is an important consideration in theme design - if these spaces all defaulted to something that matched it would mean that designs would look good without needing to tweak all these things individually as we do today cc @kjellr @beafialho

I spent some time thinking about what other instances this could be helpful with so that we could have a better visual result as default. Usually, there may be the intention of using the same padding in most theme elements (such as text in Cover Blocks, Query Loops and even Posts). But there may be cases where that's not the intention and we may end up with either very constricted or too spaced out layouts.

<img width="1370" alt="Captura de ecrã 2022-04-11, às 14 56 10" src="https://user-images.githubusercontent.com/65220155/162755169-e220da60-e6e6-4351-b67c-0acd080d7644.png">

So overall, I find using the same padding can be useful as well in navigation and galleries.

<img width="1527" alt="Captura de ecrã 2022-04-11, às 14 54 54" src="https://user-images.githubusercontent.com/65220155/162754947-3f748d75-3c31-44e3-8d50-a2a8a7202dae.png">

tellthemachines committed:
Add opt-in setting via package.json

tellthemachines committed:
Generate correct block editor styles

tellthemachines committed:
Fix tests and add prop to core theme.json

tellthemachines commented:
I've updated this PR to make the new behaviour opt-in for themes. That way it won't break existing themes or websites that may be relying on the previous root padding behaviour. 

The opt-in is a line - `"useRootVariables": true,` in the theme.json settings object. This location makes sense to me because it's possible we may add a UI toggle for this behaviour in the future. 

There's also the open question of [how far](https://github.com/WordPress/gutenberg/pull/39926#discussion_r846898166) we should take support for string values in `padding`. I don't think we need to fully resolve that in this PR, but it's a conversation that we'll need to have sooner or later 😅 

Apart from further testing and review of these changes, the PR should be ready to wrap up 🎁

tellthemachines committed:
Fix unit tests properly this time

jffng commented:
I tested this against a modified Twenty Twenty-Two, removing all the theme provided alignment CSS to get full widths with padding to work: https://github.com/WordPress/wordpress-develop/pull/2578. 

I nearly shed a tear at how it just worked. Thank you for working on this!

ramonjd committed:
Merge remote-tracking branch 'origin/trunk' into try/root-padding-fix

# Conflicts:
#	packages/edit-site/src/components/global-styles/test/use-global-styles-output.js
#	packages/edit-site/src/components/global-styles/use-global-styles-output.js

ramonjd committed:
Merge remote-tracking branch 'origin/trunk' into try/root-padding-fix

# Conflicts:
#	packages/edit-site/src/components/global-styles/test/use-global-styles-output.js
#	packages/edit-site/src/components/global-styles/use-global-styles-output.js

ramonjd committed:
Added missing doc comment for parameter "$use_root_vars"

ramonjd commented:
I just merged this branch with trunk and fixed conflicts. It's working as described save for this overflow in the post editor for `alignfull` items.

<img width="964" alt="Screen Shot 2022-04-26 at 3 44 26 pm" src="https://user-images.githubusercontent.com/6458278/165229995-350001f0-8cf3-468d-8ff1-8fea6b887344.png">

I'll take a closer look tomorrow.

scruffian commented:
I tested this with TT2 and on Archeo using [this PR](https://github.com/Automattic/themes/pull/5949). It doesn't seem to be working correctly. Full width items aren't going to the browser edge, and don't have the correct padding applied.

I see the problem in the single post template - the post content template has the "inherit default" property set, but then the CSS only targets direct children of `.wp-site-blocks`, not children of other blocks that inherit the default layout. I'm not sure if this was always an issue or if the rebase introduced it. I'm happy to help work on it if that would be useful?

ramonjd commented:
>  I'm not sure if this was always an issue or if the rebase introduced it.

Thanks for spending time on this. It's possible I borked something during the merge with trunk. It was a bit hairy due to all the compat changes for WordPress 6.0. I'll compare with a26beda804cdca7e6d3b9d4b2fa9de4e6c2daead and see if I can spot any omissions

>  I'm happy to help work on it if that would be useful

That would be **very** useful, thank you! 😄 @tellthemachines is taking a break so it'd be great to keep this one rolling.

ramonjd commented:
> It was a bit hairy due to all the compat changes for WordPress 6.0. I'll compare with https://github.com/WordPress/gutenberg/commit/a26beda804cdca7e6d3b9d4b2fa9de4e6c2daead and see if I can spot any omissions

I couldn't find anything glaring (I checked every change since the merge). 

The only spot I'm suspicious about is the root CSS here: https://github.com/WordPress/gutenberg/pull/39926/files#diff-1654d2da9097b71ac77806bbc505483b794371e8a2ff3ba152ec8610c2663ec6R438

The main challenge was merging after https://github.com/WordPress/gutenberg/commit/49e4a4f17a59bd8d76d1fc148ca0e298474f0436... so maybe something needs to be updated to take into account that commit.

jffng commented:
> Full width items aren't going to the browser edge, and don't have the correct padding applied.

I gave the latest version and can verify this on TT2. To add a little more context:

As @scruffian mentions, this PR only targets direct children of `wp-site-blocks`, i.e. 

`.wp-site-blocks > * > .alignfull` 

TT2's templates wrap the post content block inside of another group block. This is for good reason — to include a `main` element on the page, required for accessibility. This means alignfull blocks that are direct children of post content are actually two levels deep: 

`.wp-site-blocks > main.wp-group > .wp-block-post-content > .alignfull`

The current alignment CSS in TT2 gets around this by being more general with its selector:

`.wp-site-blocks .alignfull`

scruffian commented:
> TT2's templates wrap the post content block inside of another group block. This is for good reason — to include a main element on the page, required for accessibility. This means alignfull blocks that are direct children of post content are actually two levels deep:

I added a commit to address this. It would be great if more people could test to find other issues that I missed :). Feel free to revert if it breaks stuff.

jffng reviewed (approved):
> I added a commit to address this.

Tested a variety of scenarios and full width patterns in TT2, looking good to me. Would love to try this out in the wild and see how it works in practice, also safer that it's opt in under the "useRootVariables" setting.

youknowriad commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -372,7 +372,7 @@ protected function get_block_classes( $style_nodes ) {
 					$block_rules .= '.wp-site-blocks { padding-top: var(--wp--style--root--padding-top); padding-bottom: var(--wp--style--root--padding-bottom); }';
 					$block_rules .= '.wp-site-blocks > * { padding-right: var(--wp--style--root--padding-right); padding-left: var(--wp--style--root--padding-left); }';
-					$block_rules .= '.wp-site-blocks > * > .alignfull { margin-right: calc(var(--wp--style--root--padding-right) * -1); margin-left: calc(var(--wp--style--root--padding-left) * -1); }';
+					$block_rules .= '.wp-site-blocks > * .alignfull { margin-right: calc(var(--wp--style--root--padding-right) * -1); margin-left: calc(var(--wp--style--root--padding-left) * -1); padding-right: var(--wp--style--root--padding-right); padding-left: var(--wp--style--root--padding-left); width: unset; }';
```
This basically targets all align full blocks at all levels. I'm not sure that's what we want, I thought this was only about the top level ones.

scruffian replied:
The aim of the PR originally was just top level blocks and those inside post content. Blocks inside post content weren't working which is why I added this line. I know this will broaden the scope of the CSS, but this is the CSS that we have been using in all block themes, so it seems fairly safe to try.

scruffian commented on lib/compat/wordpress-6.0/theme.json:
```diff
@@ -240,6 +241,8 @@
 	"styles": {
-		"spacing": { "blockGap": "24px" }
+		"spacing": {
+			"blockGap": "24px"
```
I removed the default padding from here as it would impact all themes, which would be a problem.

scruffian commented:
I spent quite a bit of time on this today. In the end I took the CSS we currently use in block themes and then adapted it. This means we use the class names for several blocks that have different behaviours (columns, cover and group). I think it's necessary to give these specific blocks a different behaviour as we need to stop the content from touching the edges in these blocks.

I'd be grateful to hear what others think. Can this also replace https://github.com/WordPress/gutenberg/pull/39871?

ramonjd commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -362,6 +369,30 @@ protected function get_block_classes( $style_nodes ) {
+				if ( $use_root_vars ) {
+					$block_rules .= '.wp-site-blocks { padding-top: var(--wp--style--root--padding-top); padding-bottom: var(--wp--style--root--padding-bottom); }';
+
+					$block_rules .= '.wp-site-blocks,
```
I'm wondering if we need to remove the padding applied directly to `.wp-site-blocks`

The padding set in theme.json and global styles are also set in the root selector (body) via `static::to_ruleset` above.

<img width="1173" alt="Screen Shot 2022-05-02 at 1 28 50 pm" src="https://user-images.githubusercontent.com/6458278/166180927-e31fed96-f381-4019-8894-8eb12838ae76.png">

scruffian replied:
Thanks for testing. I don't think they are applied to `body`, only the variables are defined on the body. In your screenshot one section is the vertical padding and the other is the horizontal padding :)

ramonjd commented on lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php:
```diff
@@ -529,6 +560,143 @@ protected static function get_metadata_boolean( $data, $path, $default = false )
+				);
+
+				$declarations = array_merge( $declarations, $all_properties );
+
```
I don't know if it's been brought up yet, but this is a lot of bespoke code for root padding values future iterations. I'm not sure where we'd put it in future iterations. 

I'm not saying it's a blocker, just curious to learn whether we need a separate processing method for root block selectors to keep these methods uncluttered.

I'd speculate that root style CSS vars could be useful elsewhere, e.g., `--wp--style--root--block-gap`

scruffian replied:
Yeah I agree a separate method for root selectors might make this tidier.

scruffian replied:
Done!

ramonjd reviewed (commented):
Thanks a lot for picking this up and running with it @scruffian 🙇 

It's mainly working as expected for me (aside from the doubling up on padding styles in body and .wp-site-blocks).

I'd be keen to hear from @youknowriad and others on whether the additional padding-specific changes in theme json class are okay, or whether they have the potential to create barriers to future refactoring.

It seems like it addresses a big headache for theme authors, so I think the tradeoffs could be worth it.

scruffian commented:
I squashed and rebased this and moved it to https://github.com/WordPress/gutenberg/pull/40926 to keep the history here in tact. I think it's in a good place.

scruffian commented:
I removed my commits from this so the original intent is clear.

scruffian commented:
I tried combining this PR with https://github.com/WordPress/gutenberg/pull/39871/. The problem is that the way this works at the moment in trunk is that the outermost block that has "inherit default layout" has to cancel out the padding set on the root. This is challenging because the outermost block to "inherit default layout" is often not a child of ".wp-site-blocks".

If we want to keep this behaviour, I think what we need is to find a way to determine if a block is wrapped in other "inherit default layout" blocks. If it isn't then it is the outermost block and needs to have negative margins to counter the root padding.

tellthemachines commented:
Closing this PR as #42034 has an updated version of the same solution.
